### PR TITLE
implement `gt sync`

### PR DESCRIPTION
**Pull request by charlierudolph**

### CODE REVIEW

File: features/git-town-sync/folder_does_not_exist_on_main_branch/no_conflict.feature [hunk 1]
```diff
@@ -49,20 +49,3 @@ Feature: git town-sync: syncing inside a folder that doesn't exist on the main b
       | other-feature   | local and remote | other feature commit                     | file2            |
       |                 |                  | main commit                              | main_file        |
       |                 |                  | Merge branch 'main' into other-feature   |                  |
-
-
-  Scenario: undo
```
Comment (kevgo) [inline features/git-town-sync/folder_does_not_exist_on_main_branch/no_conflict.feature]: Why do you delete this scenario?

Comment (charlierudolph) [inline features/git-town-sync/folder_does_not_exist_on_main_branch/no_conflict.feature]: I don't think we need to implement undo for sync. Its not useful based on the commands it runs.

File: cmd/hack.go [hunk 2]
```diff
@@ -12,29 +12,34 @@ import (
   "github.com/spf13/cobra"
 )
 
-var abortFlag bool
-var continueFlag bool
+type HackFlags struct {
```
Comment (charlierudolph) [inline cmd/hack.go]: Most of our commands have an abort / continue but actually not all of them. Only those that have the potential to hit conflicts should really have them. `repo`, `prune-branches`, `rename-branch`, and `kill` all don't need them.

File: cmd/config/config.go [hunk 3]
```diff
@@ -9,29 +9,52 @@ import (
 )
 
 
+func CompileAncestorBranches(branchName string) (result []string) {
+  current := branchName
+  for {
+    parent := GetParentBranch(current)
+    result = append([]string{parent}, result...)
+    if parent == GetMainBranch() || IsPerennialBranch(parent) {
+      break
```
Comment (charlierudolph) [inline cmd/config/config.go]: done

File: lib/config/config.go [hunk 4]
```diff
@@ -9,29 +9,51 @@ import (
 )
 
 
+func CompileAncestorBranches(branchName string) (result []string) {
```
Comment (kevgo) [inline lib/config/config.go]: Does this method really "compile" the ancestor branches? What about naming this "CalculateAncestorBranches"?

File: lib/config/config.go [hunk 5]
```diff
@@ -56,6 +78,13 @@ func IsPerennialBranch(branchName string) bool {
 }
 
 
+func KnowsAllAncestorBranches(branchName string) bool {
+  return branchName == GetMainBranch() ||
```
Comment (kevgo) [inline lib/config/config.go]: Thoughts about creating a helper method "IsMainBranch" to make this simpler and more expressive?

File: lib/config/config.go [hunk 6]
```diff
@@ -56,6 +78,13 @@ func IsPerennialBranch(branchName string) bool {
...
+    IsPerennialBranch(branchName) ||
+    len(GetAncestorBranches(branchName)) > 0
```
Comment (kevgo) [inline lib/config/config.go]: All we seem to care about here is whether there is a valid entry for ancestor branches in the Git config. Maybe create a helper called "HasStoredAncestorBranches" to make this code more expressive?

File: lib/config/config.go (hunk 4 above)
Comment (charlierudolph) [inline lib/config/config.go]: There isn't really any calculation either. Compile has multiple definitions:

> 1. produce (something, especially a list, report, or book) by assembling information collected from other sources. "the local authority must compile a list of taxpayers"
synonyms: assemble, put together, make up, collate, compose, organize, arrange
> 2. COMPUTING (of a computer) convert (a program) into a machine-code or lower-level form in which the program can be executed.

 This is what it was in the bash src named it and I prefer it to calculate.

File: lib/config/config.go (hunk 5 above)
Comment (charlierudolph) [inline lib/config/config.go]: sounds good

File: lib/config/config.go (hunk 6 above)
Comment (charlierudolph) [inline lib/config/config.go]: sounds good

File: lib/config/config.go (hunk 4 above)
Comment (kevgo) [inline lib/config/config.go]: Okay, I'm convinced

**Comment by kevgo** [comment]
Tests are red

**Pull request merged**